src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -93,6 +108,173 @@ internal ModelPath WorksharingModelPath
+
+            return purgedMaterials;
+        }
+
+        private static HashSet<ElementId> GetNonPurgableMaterialIds(Autodesk.Revit.DB.Element type)
+        {
+            HostObjAttributes hostObj = type as HostObjAttributes;
+            HashSet<ElementId> usedMaterialIds = new HashSet<ElementId>();
+            if (hostObj != null)
+            {
+                var compundStructure = hostObj.GetCompoundStructure();
+                if (compundStructure != null)
+                {
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        usedMaterialIds.Add(compundStructure.GetMaterialId(j));
+                    }
+                }
+            }
+
+            List<ElementId> elementMaterialIds = type.GetMaterialIds(false).ToList();
+            if (elementMaterialIds.Any())
+            {
+                usedMaterialIds.UnionWith(elementMaterialIds);
+            }
+            return usedMaterialIds;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> PurgeMaterialAssets(List<Autodesk.Revit.DB.ElementId> elementIds)
+        {
+            List<ElementId> appearanceAssetIds = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(AppearanceAssetElement))
+                .ToElementIds()
+                .ToList();
+
+            List<ElementId> propertySet = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(PropertySetElement))
+                .ToElementIds()
+                .ToList();
+
+            int elementIdsCount = elementIds.Count();
+            for (int i = 0; i < elementIdsCount; i++)
+            {
+                try
+                {
+                    var material = this.InternalDocument.GetElement(elementIds[i]) as Autodesk.Revit.DB.Material;
+                    propertySet.Remove(material.ThermalAssetId);
+                    propertySet.Remove(material.StructuralAssetId);
+                    appearanceAssetIds.Remove(material.AppearanceAssetId);
+                }
+                catch (Exception)
+                {
+                    continue;
+                }
+            }
+
+            if (appearanceAssetIds.Count > 0)
+                this.InternalDocument.Delete(appearanceAssetIds);
+            if (propertySet.Count > 0)
+                this.InternalDocument.Delete(propertySet);
+
+            var purgedElementIds = new List<ElementId>(appearanceAssetIds.Concat(propertySet));
+            return purgedElementIds;
+        }
+
+        private List<ElementId> PurgeElements(bool runRecursively)
+        {
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            List<ElementId> purgeableElementIds = GetPurgeableElementIds();
+            
+            if (!runRecursively)
+            {
+                if (purgeableElementIds == null)
+                    return purgedElementIds;
```

Comment:
this is repeated both inside and after the `if`, so safe to put it before the if only once.